### DISCUSSION TITLE
Migrate react-map-gl to deck.gl and use Mapbox Style Specification

### DISCUSSION THREAD

**Discussion by rgazelot** [visgl/deck.gl#6805] [Q&A] [answered]
Hello deck.gl team and community, I'm using react-map-gl but now I have to add the deck.gl on top to benefit for some of your great features. All my layers from InteractiveMap are now inactive because I switched to StaticMap. May I ask some help to achieve my migration process:

1. I there a way to make my react-map-gl layers interactive without re-write them using a deck.gl Layer? Should I keep using InteractiveMap somehow instead of switching to StaticMap even If it's more performant?
2. I there a way to create a deck.gl Layer using all the [Mapbox Style Specification for Layers](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/)? My designer uses Mapbox Studio to design layers and I don't see how to replicate the capabilities of the Style Specification with deck.gl layers?

Thanks a lot for your help and your answers. Keep up the good work!

**Comment by Pessimistress** [comment]
https://deck.gl/docs/api-reference/mapbox/overview

**Reply by rgazelot** [reply]
Thanks @Pessimistress for the sharing. If I follow the steps, Deck.gl disables all the interactivity I had before when using InteractiveMap from react-map-gl because I switched to a StaticMap. The Deck.gl's getCursor, onHover or onClick methods don't detect any of the StaticMap child layers. How can I use my previous `interactiveLayerIds` props on `InteractiveMap` to tell Deck.gl: "look at these layers they are also interactive"?

**Reply by Pessimistress** [reply]
You can do one of two things:

1. Use deck.gl as an overlay

```js
<DeckGL getCursor={...}
  onHover={callbackOnDeckLayerHover}
  onClick={callbackOnDeckLayerClick} >
    <InteractiveMap
      interactiveLayerIds=[...]
      onHover={callbackOnMapLayerHover}
      onClick={callbackOnMapLayerClick} />
</DeckGL>
```

2. Use deck.gl as Mapbox layers

```js
const onLoad = useCallback((e) => {
  e.target.addLayer(new MapboxLayer({..., onClick: callbackOnDeckLayerClick}));
}, []);

<MapGL
  interactiveLayerIds=[...]
  onHover={callbackOnMapLayerHover}
  onClick={callbackOnMapLayerClick}
  onLoad={onLoad}
  getCursor={...} >
</MapGL>
```

Either way you need to collate the hovered object yourself in a state. Mapbox and deck.gl use independent event engines and they do not know what the other library draws to the canvas.

**Reply by rgazelot** [reply]
Thanks a lot @Pessimistress for your help. This is a try I made on codesandbox that seems to work, at least to put the first stones. I paste it here for the record in case someone else is looking for the same results:

```js
const Map = () => {
  const onMapLoad = (e) => {
    const map = e.target;
    const gl = map.painter.context.gl;

    const deck = new Deck({
      gl,
      layers: [
        new GeoJsonLayer({
          id: "deck-polygon",
          data: deckPolygonData,
          pickable: true,
          getFillColor: [255, 0, 0],
          onHover: event => console.log("hover on geojson deck layer", event),
          onClick: event => console.log("click on geojson deck layer", event)
        })
      ],
      controller: true,
      onClick: event => console.log("click on deck layer", event)
    });

    map.addLayer(new MapboxLayer({ id: "deck-polygon", deck }));
  };

  return (
    <MapGL
      onLoad={onMapLoad}
      width="100%"
      height="100%"
      initialViewState={INITIAL_VIEW_STATE}
      mapStyle="mapbox://styles/mapbox/light-v9"
      mapboxAccessToken={MAPBOX_TOKEN}
      interactiveLayerIds={["polygon-layer"]}
      onClick={(event) => console.log("click on react-map-gl layers", event.features)}
    >
      <PolygonLayer />
    </MapGL>
  );
};

export default Map;
```

**Comment by rgazelot** [comment, marked as answer]
To answer the Mapbox Style Specification question, according to https://github.com/visgl/deck.gl/discussions/6720 Deck.gl doesn't support the Style Spec on any of its layers.

**Reply by chrisgervang** [reply]
@kylebarron and @ibgreen at Unfolded started an effort to make basemap layers that were compatible with the Style Spec. It needs a lot more work before it's ready, but they were encouraging myself and others who expressed interest in contributing. https://github.com/UnfoldedInc/basemap-layers
